core/platform/email/mailgun_email_services.py

```diff
@@ -14,138 +14,116 @@
-            entity.
-        bcc_admin: bool. Whether to bcc feconf.ADMIN_EMAIL_ADDRESS on the email.
-        reply_to_id: str. The unique id of the sender.
-
-    Raises:
-        Exception: if the configuration in feconf.py forbids emails from being
-            sent.
-        Exception: if mailgun api key is not stored in feconf.MAILGUN_API_KEY.
-        Exception: if mailgun domain name is not stored in
-            feconf.MAILGUN_DOMAIN_NAME.
-            (and possibly other exceptions, due to mail.send_mail() failures)
-    """
-    if not feconf.CAN_SEND_EMAILS:
-        raise Exception('This app cannot send emails to users.')
-
-    data = {
-        'from': sender_email,
-        'to': recipient_email,
-        'subject': subject.encode(encoding='utf-8'),
-        'text': plaintext_body.encode(encoding='utf-8'),
-        'html': html_body.encode(encoding='utf-8')
-    }
-
-    if bcc_admin:
-        data['bcc'] = feconf.ADMIN_EMAIL_ADDRESS
-
-    if reply_to_id:
-        reply_to = gae_email_services.get_incoming_email_address(reply_to_id)
-        data['h:Reply-To'] = reply_to
-
-    post_to_mailgun(data)
-
-
-def send_bulk_mail(
-        sender_email, recipient_emails, subject, plaintext_body, html_body):
-    """Sends an email using mailgun api.
-
-    In general this function should only be called from
-    email_manager._send_email().
-
-    Args:
-        sender_email: str. the email address of the sender. This should be in
-            the form 'SENDER_NAME <SENDER_EMAIL_ADDRESS>'.
-        recipient_emails: list(str). list of the email addresses of recipients.
-        subject: str. The subject line of the email.
-        plaintext_body: str. The plaintext body of the email.
-        html_body: str. The HTML body of the email. Must fit in a datastore
-            entity.
-
-    Raises:
-        Exception: if the configuration in feconf.py forbids emails from being
-            sent.
-        Exception: if mailgun api key is not stored in feconf.MAILGUN_API_KEY.
-        Exception: if mailgun domain name is not stored in
-            feconf.MAILGUN_DOMAIN_NAME.
-            (and possibly other exceptions, due to mail.send_mail() failures)
-    """
-    if not feconf.CAN_SEND_EMAILS:
-        raise Exception('This app cannot send emails to users.')
-
     # To send bulk emails we pass list of recipients in 'to' paarameter of
     # post data. Maximum limit of recipients per request is 1000.
     # For more detail check following link:
     # https://documentation.mailgun.com/user_manual.html#batch-sending
-    recipient_email_sets = [
+    recipient_email_lists = [
         recipient_emails[i:i + 1000]
         for i in python_utils.RANGE(0, len(recipient_emails), 1000)]
+    for email_list in recipient_email_lists:
+        data = {
+            'from': sender_email,
+            'subject': subject,
+            'text': plaintext_body,
+            'html': html_body
+        }
+
+        if len(email_list) == 1:
```

Comment:
Here and below, do you actually need the branching?

If you do, could also simplify to `data['to'] = ... if len(email_list) == 1 else ...`, and ditto below.